Error when compressing files of specific size with -B4 on the command line

**Describe the bug**

Error when compressing files of a size 65533 to 65535 bytes when using "lz4 -B4"

**To Reproduce**
On the command line:

[lz4_err]$ yes | head -c 65533 | lz4 -B4 >file_cout
using blocks of size 64 KB 
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall 

[lz4_err]$ echo $?
41

[lz4_err]$ ls -l
total 0
-rw-r--r--. 1 marcmassar marcmassar 0 Mar  6 22:28 file_cout



**System (please complete the following information):**
 - OS: tested on rel8, fedora. 
 - Version [e.g. 22]
 
[lz4_err]$ lz4 --version
*** LZ4 command line interface 64-bits v1.9.4, by Yann Collet ***

Hi @MarcMassar 
Thank you for the report.

First of all, I've confirmed your issue with `v1.9.4`.  I believe it's our bug.
But I also don't confirm this issue with the recent (not-released) version of lz4 repo.
It means that this bug had been (accidentally) fixed.

Therefore, if you need the fix now, please pull the recent version lz4 from `dev` branch of this repo.
Actual commit hash of the fix is e58a06d1f9df02c85bfa15abcfd97118f9d355d6 (part of #1336).

----

We can find the first *good* commit with [this method by Michael Wolf](https://stackoverflow.com/a/17153598).

<details><summary>git bisect log</summary>

```
/lz4-issue-1374$ git bisect start --term-new=fixed --term-old=unfixed
/lz4-issue-1374$ git bisect fixed dev
/lz4-issue-1374$ git bisect unfixed 5ff839680134437dbf4678f3d0c7b371d84f4964


Bisecting: a merge base must be tested
[cfd6ab32522280079c2e6d3ea995f172b9ae0312] update NEWS for v1.9.4
/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
cfd6ab3 (HEAD) update NEWS for v1.9.4
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 230 revisions left to test after this (roughly 8 steps)
[2b755f4594b7e9bf8247394c3a2474621b138412] Merge pull request #1302 from lz4/dependabot/github_actions/github/codeql-action-2.22.4


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
2b755f4 (HEAD) Merge pull request #1302 from lz4/dependabot/github_actions/github/codeql-action-2.22.4
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 115 revisions left to test after this (roughly 7 steps)
[db7495d08b52ef9e51c4233a9b3c82d0739d2ec0] fix another C90 pedantic warning


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
db7495d (HEAD) fix another C90 pedantic warning
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
==> building with multithreading support
make[1]: Leaving directory '/lz4-issue-1374/programs'
lz4 build completed
using blocks of size 64 KB
/lz4-issue-1374$ git bisect fixed
Bisecting: 57 revisions left to test after this (roughly 6 steps)
[5c9e349c4a7ec8f6449f37f26e24882586b35a82] changed test name cpp->cxx


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
5c9e349 (HEAD) changed test name cpp->cxx
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 28 revisions left to test after this (roughly 5 steps)
[67648edb8c5aba25aa94836215f51e15225e8b81] Merge pull request #1331 from lz4/man_mt


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
67648ed (HEAD) Merge pull request #1331 from lz4/man_mt
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 14 revisions left to test after this (roughly 4 steps)
[e58a06d1f9df02c85bfa15abcfd97118f9d355d6] fixed smaller block sizes


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
e58a06d (HEAD) fixed smaller block sizes
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
/lz4-issue-1374$ git bisect fixed
Bisecting: 6 revisions left to test after this (roughly 3 steps)
[127fa5789502e112e2824c004168c546a45bf401] control over nbThreads


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
127fa57 (HEAD) control over nbThreads
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
lz4io.c: In function ‘LZ4IO_setNbWorkers’:
lz4io.c:199:1: warning: control reaches end of non-void function [-Wreturn-type]
  199 | }
      | ^
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[3acabda180389bfc5ea095c6ffcb1e71562ea7ea] multithreading works with lz4f format, but


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
3acabda (HEAD) multithreading works with lz4f format, but
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 1 revision left to test after this (roughly 1 step)
[d4c21a4dd0aafee521d4da0a2ebadd17a427b200] fixed stdin support


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
d4c21a4 (HEAD) fixed stdin support
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[95c2bc7f1f6dac251d95853cde461f55e9562a16] fix legacy format compression


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
95c2bc7 (HEAD) fix legacy format compression
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
e58a06d1f9df02c85bfa15abcfd97118f9d355d6 is the first fixed commit
commit e58a06d1f9df02c85bfa15abcfd97118f9d355d6
Author: Yann Collet <cyan@fb.com>
Date:   Wed Dec 27 15:59:53 2023 -0800

    fixed smaller block sizes

 lib/lz4frame.c   |  2 +-
 programs/lz4io.c | 71 +++++++++++++++++++++++++++++++++-----------------------
 2 files changed, 43 insertions(+), 30 deletions(-)
```

</details>


<details><summary>Confirmation</summary>

```
cd
git clone https://github.com/lz4/lz4.git lz4-issue-1374-check
cd lz4-issue-1374-check

git log -3 --pretty=format:"%h %ad %s" e58a06d
# e58a06d Wed Dec 27 15:59:53 2023 -0800 fixed smaller block sizes
# 95c2bc7 Wed Dec 27 15:22:53 2023 -0800 fix legacy format compression
# d4c21a4 Wed Dec 27 14:17:02 2023 -0800 fixed stdin support

git checkout e58a06d
git log -1 && make && ./lz4 --version && ( yes | head -c 65533 | ./lz4 -B4 >file_cout )
# OK

git checkout 95c2bc7
git log -1 && make && ./lz4 --version && ( yes | head -c 65533 | ./lz4 -B4 >file_cout )
# NG

git checkout v1.9.4
git log -1 && make && ./lz4 --version && ( yes | head -c 65533 | ./lz4 -B4 >file_cout )
# NG
```
</details>

--- a/programs/lz4cli.c
+++ b/programs/lz4cli.c
@@ -88,7 +88,9 @@ static unsigned displayLevel = 2;   /* 0 : no display ; 1: errors only ; 2 : dow
 /*-************************************
 *  Errors and Messages
 ***************************************/
-#define DEBUG 0
+#ifndef DEBUG
+#  define DEBUG 0
+#endif
 #define DEBUGOUTPUT(...) do { if (DEBUG) DISPLAY(__VA_ARGS__); } while (0)
 #define END_PROCESS(error, ...)                                   \
 do {                                                              \

--- a/programs/lz4io.c
+++ b/programs/lz4io.c
@@ -965,6 +965,7 @@ typedef struct {
     void*  dstBuffer;
     size_t dstBufferSize;
     LZ4F_compressionContext_t ctx;
+    LZ4F_preferences_t preparedPrefs;
     LZ4F_CDict* cdict;
     TPOOL_ctx* tpool;
     TPOOL_ctx* wpool; /* writer thread */
@@ -1044,37 +1045,48 @@ static void* LZ4IO_createDict(size_t* dictSize, const char* dictFilename)
     return dictBuf;
 }
 
-static LZ4F_CDict* LZ4IO_createCDict(const LZ4IO_prefs_t* prefs)
+static LZ4F_CDict* LZ4IO_createCDict(const LZ4IO_prefs_t* io_prefs)
 {
     size_t dictionarySize;
     void* dictionaryBuffer;
     LZ4F_CDict* cdict;
-    if (!prefs->useDictionary) return NULL;
-    dictionaryBuffer = LZ4IO_createDict(&dictionarySize, prefs->dictionaryFilename);
+    if (!io_prefs->useDictionary) return NULL;
+    dictionaryBuffer = LZ4IO_createDict(&dictionarySize, io_prefs->dictionaryFilename);
     if (!dictionaryBuffer) END_PROCESS(29, "Dictionary error : could not create dictionary");
     cdict = LZ4F_createCDict(dictionaryBuffer, dictionarySize);
     free(dictionaryBuffer);
     return cdict;
 }
 
-static cRess_t LZ4IO_createCResources(const LZ4IO_prefs_t* prefs)
+static cRess_t LZ4IO_createCResources(const LZ4IO_prefs_t* io_prefs)
 {
     const size_t chunkSize = 4 MB;
     cRess_t ress;
+    memset(&ress, 0, sizeof(ress));
 
-    LZ4F_errorCode_t const errorCode = LZ4F_createCompressionContext(&(ress.ctx), LZ4F_VERSION);
-    if (LZ4F_isError(errorCode))
-        END_PROCESS(30, "Allocation error : can't create LZ4F context : %s", LZ4F_getErrorName(errorCode));
+    /* set compression advanced parameters */
+    ress.preparedPrefs.autoFlush = 1;
+    ress.preparedPrefs.frameInfo.blockMode = (LZ4F_blockMode_t)io_prefs->blockIndependence;
+    ress.preparedPrefs.frameInfo.blockSizeID = (LZ4F_blockSizeID_t)io_prefs->blockSizeId;
+    ress.preparedPrefs.frameInfo.blockChecksumFlag = (LZ4F_blockChecksum_t)io_prefs->blockChecksum;
+    ress.preparedPrefs.frameInfo.contentChecksumFlag = (LZ4F_contentChecksum_t)io_prefs->streamChecksum;
+    ress.preparedPrefs.favorDecSpeed = io_prefs->favorDecSpeed;
 
-    /* Allocate Memory */
+    /* Allocate compression state */
+    {   LZ4F_errorCode_t const errorCode = LZ4F_createCompressionContext(&(ress.ctx), LZ4F_VERSION);
+        if (LZ4F_isError(errorCode))
+            END_PROCESS(30, "Allocation error : can't create LZ4F context : %s", LZ4F_getErrorName(errorCode));
+    }
+
+    /* Allocate Buffers */
     ress.srcBuffer = malloc(chunkSize);
     ress.srcBufferSize = chunkSize;
-    ress.dstBufferSize = LZ4F_compressFrameBound(chunkSize, NULL);   /* cover worst case */
+    ress.dstBufferSize = LZ4F_compressFrameBound(chunkSize, &ress.preparedPrefs);
     ress.dstBuffer = malloc(ress.dstBufferSize);
     if (!ress.srcBuffer || !ress.dstBuffer)
         END_PROCESS(31, "Allocation error : can't allocate buffers");
 
-    ress.cdict = LZ4IO_createCDict(prefs);
+    ress.cdict = LZ4IO_createCDict(io_prefs);
 
     /* will be created it needed */
     ress.tpool = NULL;
@@ -1143,16 +1155,10 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
     if (srcFile == NULL) return 1;
     dstFile = LZ4IO_openDstFile(dstFileName, io_prefs);
     if (dstFile == NULL) { fclose(srcFile); return 1; }
-    memset(&prefs, 0, sizeof(prefs));
 
-    /* Set compression parameters */
-    prefs.autoFlush = 1;
+    /* Adjust compression parameters */
+    prefs = ress.preparedPrefs;
     prefs.compressionLevel = compressionLevel;
-    prefs.frameInfo.blockMode = (LZ4F_blockMode_t)io_prefs->blockIndependence;
-    prefs.frameInfo.blockSizeID = (LZ4F_blockSizeID_t)io_prefs->blockSizeId;
-    prefs.frameInfo.blockChecksumFlag = (LZ4F_blockChecksum_t)io_prefs->blockChecksum;
-    prefs.frameInfo.contentChecksumFlag = (LZ4F_contentChecksum_t)io_prefs->streamChecksum;
-    prefs.favorDecSpeed = io_prefs->favorDecSpeed;
     if (io_prefs->contentSizeFlag) {
       U64 const fileSize = UTIL_getOpenFileSize(srcFile);
       prefs.frameInfo.contentSize = fileSize;   /* == 0 if input == stdin */
@@ -1359,14 +1365,9 @@ LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
     if (dstFile == NULL) { fclose(srcFile); return 1; }
     memset(&prefs, 0, sizeof(prefs));
 
-    /* Set compression parameters */
-    prefs.autoFlush = 1;
+    /* Adjust compression parameters */
+    prefs = ress.preparedPrefs;
     prefs.compressionLevel = compressionLevel;
-    prefs.frameInfo.blockMode = (LZ4F_blockMode_t)io_prefs->blockIndependence;
-    prefs.frameInfo.blockSizeID = (LZ4F_blockSizeID_t)io_prefs->blockSizeId;
-    prefs.frameInfo.blockChecksumFlag = (LZ4F_blockChecksum_t)io_prefs->blockChecksum;
-    prefs.frameInfo.contentChecksumFlag = (LZ4F_contentChecksum_t)io_prefs->streamChecksum;
-    prefs.favorDecSpeed = io_prefs->favorDecSpeed;
     if (io_prefs->contentSizeFlag) {
       U64 const fileSize = UTIL_getOpenFileSize(srcFile);
       prefs.frameInfo.contentSize = fileSize;   /* == 0 if input == stdin */

--- a/tests/test-lz4-basic.sh
+++ b/tests/test-lz4-basic.sh
@@ -77,3 +77,5 @@ diff -q $FPREFIX-test $FPREFIX-test2
 # Multithreading commands
 datagen -g16M | lz4 -T2 | lz4 -t
 datagen -g16M | lz4 --threads=2 | lz4 -t
+# bug #1374
+datagen -g4194302 | lz4 -B4 -c > $FPREFIX-test3

